Composer/packages/lib/shared/src/schemaUtils/dereference.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { JSONSchema7, JSONSchema7Definition } from 'json-schema';
+import get from 'lodash/get';
+import memoize from 'lodash/memoize';
+
+type SchemaDefinitions = {
+  [key: string]: JSONSchema7Definition;
+};
+
+const CIRCULAR_REFS = [
+  'Microsoft.IDialog',
+  'Microsoft.IRecognizer',
+  'Microsoft.ILanguageGenerator',
+  'Microsoft.ITriggerSelector',
+  'Microsoft.AdaptiveDialog',
+];
+
+const definitionCache = new Map();
+
+const isCircular = memoize((def: string) => CIRCULAR_REFS.some(kind => def.includes(kind)));
+
+function getRef(ref: string, definitions: SchemaDefinitions) {
+  const defName = ref.replace('#/definitions/', '');
+  let def = definitionCache.get(defName) || get(definitions, defName.split('/'));
+
+  if (!def) {
+    throw new Error(`Definition not found for ${defName}`);
+  }
+
+  if (!isCircular(defName)) {
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    def = dereference(def, definitions);
+  }
+  definitionCache.set(defName, def);
+
+  return def;
+}
+
+export function dereference<S extends JSONSchema7 | JSONSchema7[] | JSONSchema7Definition>(
+  schema: S,
+  definitions: SchemaDefinitions
+): S extends JSONSchema7[] ? JSONSchema7[] : S extends JSONSchema7 ? JSONSchema7 : JSONSchema7Definition {
+  // properties, oneOf, allOf, anyOf, items, additionalProperties
+  if (Array.isArray(schema)) {
+    const arraySchema: JSONSchema7[] = [];
+
+    (schema as JSONSchema7[]).forEach(s => {
+      arraySchema.push(dereference(s, definitions));
+    });
+
+    return arraySchema as never;
```

Comment:
I'm confused by this casting things to never. Why do we need it when the return type of this function already has logic to say what return types should be for the given realizations of S?